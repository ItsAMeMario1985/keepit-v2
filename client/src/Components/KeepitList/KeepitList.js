import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { useHistory } from 'react-router-dom'
import { StarIcon } from '../Icons'
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner'
import ImageNotFoundImg from '../../Assets/image-not-found.png'
import { v4 as uuidv4 } from 'uuid'

KeepitList.propTypes = {
  keepits: PropTypes.array.isRequired,
  className: PropTypes.string,
}

export default function KeepitList({ keepits, className }) {
  const history = useHistory()

  if (keepits.length === 0) {
    return <LoadingSpinner />
  }

  function gotoDetail(keepit) {
    history.push('/detail/' + keepit._id, { keepit: keepit })
  }

  return (
    <StyledUl className={className}>
      {keepits &&
        keepits.map((keepit, index) => (
          <li key={keepit._id}>
            <StyledImg
              key={keepit._id}
              src={
                keepit.images.length > 0
                  ? keepit.images[0].path.replace('.webp', '_thumb.webp')
                  : ImageNotFoundImg
              }
              alt=""
              onClick={() => gotoDetail(keepit)}
              data-testid="TestKeepitImg"
            ></StyledImg>
            <StyledStarRating key={keepit._id + 'rate'}>
              {[...Array(keepit.rating)].map((index) => (
                <StarIcon
                  key={uuidv4()}
                  width="5"
                  fill="var(--color-primary)"
                ></StarIcon>
              ))}
            </StyledStarRating>
          </li>
        ))}
    </StyledUl>
  )
}

const StyledStarRating = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  border-top-right-radius: 6px;
  font-weight: 600;
  bottom: 1px;
  left: 1px;
  position: absolute;
  height: 12px;
  z-index: 1;
  background-color: #00000059;
  padding-right: 5px;
  height: 15px;
  color: white;
  text-align: center;

  svg {
    height: 12px;
    margin-left: 3px;
  }
`
const StyledImg = styled.img`
  max-height: 100%;
  min-width: 100%;
  object-fit: cover;
  vertical-align: bottom;
  border: 1px dotted #e3e3e3;
  border-radius: 2px;
  cursor: pointer;
`
const StyledUl = styled.ul`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  align-items: flex-start;
  align-content: flex-start;

  li {
    height: 20vh;
    flex-grow: 4;
    position: relative;
    align-self: flex-start;

    &:last-child {
      flex-grow: 1;
      margin-right: auto;
    }
  }
`
