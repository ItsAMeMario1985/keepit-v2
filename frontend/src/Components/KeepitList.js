import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import loading from '../Assets/loading.gif'
import { useHistory, useLocation } from 'react-router-dom'

export default function KeepitList({ keepits }) {
  const history = useHistory()

  const StyledLoading = styled.div`
    text-align: center;
    font-size: 12px;
  `

  if (keepits.length === 0) {
    return (
      <StyledLoading>
        <img width="20" src={loading}></img>
      </StyledLoading>
    )
  }

  function gotoDetail(keepit) {
    console.log('redirect to', keepit)
    history.push('/detail/' + keepit.id, { keepit: keepit })
  }

  return (
    <StyledUl>
      {keepits.map((keepit, index) => (
        <StyledLi>
          <StyledImg
            src={'http://keepit-be.local/' + keepit.images[0]}
            alt=""
            key={keepit.id}
            onClick={() => gotoDetail(keepit)}
          ></StyledImg>
        </StyledLi>
      ))}
    </StyledUl>
  )
}
const StyledImg = styled.img`
  max-height: 100%;
  min-width: 100%;
  object-fit: cover;
  vertical-align: bottom;
  border: 1px dotted #e3e3e3;
  border-radius: 2px;
  cursor: pointer;
`
const StyledUl = styled.ul`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  flex-basis: | auto;
`
const StyledLi = styled.li`
  height: 12vh;
  flex-grow: 4;

  &:last-child {
    flex-grow: 0;
    margin-right: auto;
  }
`
