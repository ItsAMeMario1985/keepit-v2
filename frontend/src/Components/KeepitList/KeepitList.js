import PropTypes from 'prop-types'
import styled from 'styled-components/macro'
import { useHistory } from 'react-router-dom'
import { StarIcon } from '../Icons'
import LoadingSpinner from '../LoadingSpinner/LoadingSpinner'

export default function KeepitList({ keepits, className }) {
  const history = useHistory()
  const apiBaseUrl = process.env.REACT_APP_API_BASE_URL

  if (keepits.length === 0) {
    return <LoadingSpinner />
  }

  function gotoDetail(keepit) {
    console.log('redirect to', keepit)
    history.push('/detail/' + keepit.id, { keepit: keepit })
  }

  return (
    <StyledUl className={className}>
      {keepits.map((keepit, index) => (
        <StyledLi key={index}>
          <StyledImg
            src={apiBaseUrl + '/' + keepit.images[0]}
            alt=""
            onClick={() => gotoDetail(keepit)}
          ></StyledImg>
          <StyledStarRating>
            {[...Array(keepit.rated)].map(() => (
              <StarIcon width="5" fill="var(--color-primary)"></StarIcon>
            ))}
          </StyledStarRating>
        </StyledLi>
      ))}
    </StyledUl>
  )
}

const StyledStarRating = styled.div`
  display: flex;
  flex-direction: row;
  justify-content: flex-start;
  align-items: center;
  border-top-right-radius: 6px;
  font-weight: 600;
  bottom: 1px;
  left: 1px;
  position: absolute;
  height: 12px;
  z-index: 1;
  background-color: #00000059;
  padding-right: 5px;
  height: 15px;
  color: white;
  text-align: center;

  svg {
    height: 12px;
    margin-left: 3px;
  }
`
const StyledImg = styled.img`
  max-height: 100%;
  min-width: 100%;
  object-fit: cover;
  vertical-align: bottom;
  border: 1px dotted #e3e3e3;
  border-radius: 2px;
  cursor: pointer;
`
const StyledUl = styled.ul`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: flex-start;
  align-items: flex-start;
  align-content: flex-start;
`
const StyledLi = styled.li`
  height: 20vh;
  flex-grow: 4;
  position: relative;
  align-self: flex-start;

  &:last-child {
    flex-grow: 1;
    margin-right: auto;
  }
`
