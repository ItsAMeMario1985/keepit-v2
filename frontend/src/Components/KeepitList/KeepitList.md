```jsx
<KeepitList
  keepits={[
    {
      rated: 4,
      city: 'Hamburg',
      country: 'Germany',
      date: {
        date: '2020-12-29 00:00:00.000000',
        timezone_type: 3,
        timezone: 'UTC',
      },
      id: 187,
      images: [
        'https://www.test.de/images/svgicons/test.de-logo;v30116961.svg',
      ],
      tags: [
        {
          value: 'Beer',
          isCustom: false,
        },
        {
          value: 'Alcohol',
          isCustom: false,
        },
      ],
    },
    {
      rated: 4,
      city: 'Hamburg',
      country: 'Germany',
      date: {
        date: '2020-12-29 00:00:00.000000',
        timezone_type: 3,
        timezone: 'UTC',
      },
      id: 188,
      images: [
        'https://www.test.de/images/svgicons/test.de-logo;v30116961.svg',
      ],
      tags: [
        {
          value: 'Beer',
          isCustom: false,
        },
        {
          value: 'Alcohol',
          isCustom: false,
        },
      ],
    },
  ]}
/>
```
