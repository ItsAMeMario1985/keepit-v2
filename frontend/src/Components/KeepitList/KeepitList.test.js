import { render } from '@testing-library/react'
import KeepitList from './KeepitList'
import testImage from '../../Assets/test.png'

describe('KeepitList', () => {
  const keepits = [
    {
      rated: 4,
      city: 'Hamburg',
      country: 'Germany',
      date: {
        date: '2020-12-29 00:00:00.000000',
        timezone_type: 3,
        timezone: 'UTC',
      },
      id: 187,
      images: [testImage],
      tags: [
        {
          value: 'Beer',
          isCustom: false,
        },
        {
          value: 'Alcohol',
          isCustom: false,
        },
      ],
    },
  ]

  it('renders correctly', () => {
    const { container } = render(<KeepitList keepits={keepits} />)
    expect(container.firstChild).toMatchSnapshot()
  })

  // it('image shown', () => {
  //   const { container, getByTestId, getByText } = render(
  //     <KeepitList keepits={keepits} />
  //   )
  //   const ImgContainer = getByTestId('keepitImg')
  //   expect(ImgContainer).toHaveStyle(`max-height: 100%;`)
  //   expect(ImgContainer).toHaveAttribute('src', 'undefined/' + testImage)
  // })
})

/*

const data = [
  {
    "rated": 4,
    "city": "Hamburg",
    "country": "Germany",
    "date": {
      "date": "2020-12-29 00:00:00.000000",
      "timezone_type": 3,
      "timezone": "UTC"
    },
    "id": 187,
    "images": [
      "uploads/5feb0c39a9339.jpeg"
    ],
    "tags": [
      {
        "value": "Beer",
        "isCustom": false
      },
      {
        "value": "Alcohol",
        "isCustom": false
      }
    ]
  }
]

*/
