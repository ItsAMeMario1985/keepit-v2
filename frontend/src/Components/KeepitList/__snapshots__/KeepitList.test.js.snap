// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`KeepitList renders correctly 1`] = `
<ul
  class="KeepitList__StyledUl-sc-1db6amq-2 nyXXF"
>
  <li
    class="KeepitList__StyledLi-sc-1db6amq-3 eWnMnb"
  >
    <img
      alt=""
      class="KeepitList__StyledImg-sc-1db6amq-1 bA-Drmk"
      src="undefined/test.png"
    />
    <div
      class="KeepitList__StyledStarRating-sc-1db6amq-0 iDCVPP"
    >
      <svg
        fill="var(--color-primary)"
        width="5"
      >
        star.svg
      </svg>
      <svg
        fill="var(--color-primary)"
        width="5"
      >
        star.svg
      </svg>
      <svg
        fill="var(--color-primary)"
        width="5"
      >
        star.svg
      </svg>
      <svg
        fill="var(--color-primary)"
        width="5"
      >
        star.svg
      </svg>
    </div>
  </li>
</ul>
`;
